pydata_sphinx_theme/__init__.py

```diff
@@ -4,76 +4,121 @@
-        nav = [docutils_node_to_jinja(child, only_pages=True) for child in toc_items]
-
-        return nav
-
-    def get_page_toc_object():
-        """Return a list of within-page TOC links that can be accessed from Jinja."""
-        self_toc = TocTree(app.env).get_toc_for(pagename, app.builder)
-
-        try:
-            # If there's only one child, assume we have a single "title" as top header
-            # so start the TOC at the first item's children (AKA, level 2 headers)
-            if len(self_toc.children) == 1:
-                nav = docutils_node_to_jinja(self_toc.children[0]).get("children", [])
+        toc_sphinx = context["toctree"](**kwargs)
+        soup = bs(toc_sphinx, "html.parser")
+
+        # pair "current" with "active" since that's what we use w/ bootstrap
+        for li in soup("li", {"class": "current"}):
+            li["class"].append("active")
+
+        if kind == "navbar":
+            # Add CSS for bootstrap
+            for li in soup("li"):
+                li["class"].append("nav-item")
+                li.find("a")["class"].append("nav-link")
+            out = "\n".join([ii.prettify() for ii in soup.find_all("li")])
+
+        elif kind == "sidebar":
+            # Remove sidebar links to sub-headers on the page
+            for li in soup.select("li.current ul li"):
+                # Remove
+                if li.find("a"):
+                    href = li.find("a")["href"]
+                    if "#" in href and href != "#":
+                        li.decompose()
+
+            # Join all the top-level `li`s together for display
+            current_lis = soup.select("li.current.toctree-l1 li.toctree-l2")
+            out = "\n".join([ii.prettify() for ii in current_lis])
+
+        elif kind == "raw":
+            out = soup
+
+        return out
+
+    def get_page_toc_object(kind="html"):
+        """Return the within-page TOC links in HTML."""
+
+        if "toc" not in context:
+            return ""
+
+        soup = bs(context["toc"], "html.parser")
+
+        # Add toc-hN classes
+        def add_header_level_recursive(ul, level):
+            for li in ul("li", recursive=False):
+                li["class"] = li.get("class", []) + [f"toc-h{level}"]
+                ul = li.find("ul", recursive=False)
+                if ul:
+                    add_header_level_recursive(ul, level + 1)
+
+        add_header_level_recursive(soup.find("ul"), 1)
+
+        # Add in CSS classes for bootstrap
+        for ul in soup("ul"):
+            ul["class"] = ul.get("class", []) + ["nav", "section-nav", "flex-column"]
+
+        for i in range(1, context["theme_show_toc_level"] + 1):
+            for li in soup.select("li.toc-h{}".format(i + 1)):
+                ul = li.find_parent("ul")
+                classes = ul.get("class", [])
+                if "visible" not in classes:
+                    ul["class"] = classes + ["visible"]
```

Comment:
@choldgraf this is "a" way that I found to get `show_toc_level` working, but I find it king of hacky ..

The problem with the basic 

```
        for ul in soup("ul"):
            ul["class"] = ...
```

loop above this loop is that this is a flat iterator, and so you don't know how "deep" you are for each `ul`. 

This "workaround" uses the `toc-hN` class names we added above, to find the appropriate levels of `ul` for which to add the class. I didn't directly find a way with beautifulsoup to only search a certain number of levels down.

Reply:
this seems reasonable to me :-)

Reply:
Hmm, not sure why I missed that earlier. But of course the code that adds those `toc-hN` classes (the classes that I was then using here to navigate the soup as a workaround), *is already* recursively iterating through the lists keeping track of the level ... :) 
So simplified this